src/HOCON/Impl/HoconObject.cs

```diff
@@ -29,223 +31,315 @@ namespace Hocon
+            }
+            return true;
         }
 
         /// <summary>
         /// Retrieves the value associated with the supplied key.
         /// If the supplied key is not found, then one is created
         /// with a blank value.
         /// </summary>
-        /// <param name="key">The key associated with the value to retrieve.</param>
+        /// <param name="path">The path associated with the value to retrieve.</param>
         /// <returns>The value associated with the supplied key.</returns>
-        public HoconValue GetOrCreateKey(string key)
+        private HoconField GetOrCreateKey(HoconPath path)
         {
-            if (Items.ContainsKey(key))
+            if (TryGetValue(path.Key, out var child))
+                return child;
+
+            child = new HoconField(path, this);
+            Add(path.Key, child);
+            return child;
+        }
+
+        internal List<HoconField> TraversePath(HoconPath path)
+        {
+            var result = new List<HoconField>();
+            var pathLength = 1;
+            var currentObject = this;
+            while (true)
             {
-                return Items[key];
+                var child = currentObject.GetOrCreateKey(new HoconPath(path.GetRange(0, pathLength)));
+                result.Add(child);
+
+                pathLength++;
+                if (pathLength > path.Count)
+                    return result;
+
+                child.EnsureFieldIsObject();
+                currentObject = child.GetObject();
             }
-            var child = new HoconValue();
-            Items.Add(key, child);
-            return child;
         }
 
         /// <summary>
         /// Returns a HOCON string representation of this element.
         /// </summary>
         /// <returns>A HOCON string representation of this element.</returns>
         public override string ToString()
+            => ToString(0, 2);
+
+        /// <inheritdoc />
+        public string ToString(int indent, int indentSize)
         {
-            return ToString(0);
+            var i = new string(' ', indent * indentSize);
+            var list = new List<string>();
+            foreach (var field in this)
+            {
+                list.Add($"{i}{field.Key} : {field.Value.ToString(indent + 1, indentSize)}");
+            }
+            return string.Join($",{Environment.NewLine}", list);
         }
 
-        /// <summary>
-        /// Returns a HOCON string representation of this element.
-        /// </summary>
-        /// <param name="indent">The number of spaces to indent the string.</param>
-        /// <returns>A HOCON string representation of this element.</returns>
-        public string ToString(int indent)
+        public virtual void Merge(HoconObject other)
         {
-            var i = new string(' ', indent*2);
-            var sb = new StringBuilder();
-            foreach (var kvp in Items)
+            var keys = other.Keys.ToArray();
```

Comment:
Why `ToArray`? It looks like you're allocating array only for sake of enumeration.

Reply:
Because it can change the enumerated Dictionary, it will throw an exception if i didn't make a copy first.